File: src/main/java/net/finance/entity/BudgetCategories.java
package net.finance.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.*;

import javax.persistence.*;
import java.io.Serializable;
import java.math.BigDecimal;

@Entity
@Table(name = "budget_categories")
@Getter
@Setter
@EqualsAndHashCode
@NoArgsConstructor
public class BudgetCategories implements Serializable {

	/**
	 *
	 */
	private static final long serialVersionUID = -7040527905669918246L;
	@Id
	@ManyToOne
	@JoinColumn(name = "budget_id", insertable = false, updatable = false)
	@JsonIgnore
	private Budget budget;
	@Id
	@ManyToOne
	@JoinColumn(name = "category_id", insertable = false, updatable = false)
	private Category category;
	@NonNull
	@Column(name = "value")
	private BigDecimal value;

	public BudgetCategories(Category category, BigDecimal value) {
		this.category = category;
		this.value = value;
	}

}

File: src/main/kotlin/com/finances/entity/BudgetCategories.kt
package com.finances.entity

import java.io.Serializable
import java.math.BigDecimal
import javax.persistence.Column
import javax.persistence.Embeddable
import javax.persistence.EmbeddedId
import javax.persistence.Entity
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne
import javax.persistence.PrePersist
import javax.persistence.Table

@Entity
@Table(name = "budget_categories")
class BudgetCategories(
    @ManyToOne
    @JoinColumn(name = "category_id", insertable = false, updatable = false)
    var category: Category,
    @ManyToOne
    @JoinColumn(name = "budget_id", insertable = false, updatable = false)
    var budget: Budget,
    @EmbeddedId
    var budgetCategoriesId: BudgetCategoriesId = BudgetCategoriesId(category.id, budget.id),
    var value: BigDecimal = BigDecimal.ZERO,
) : Serializable {

    @PrePersist
    fun prePersist() {
        this.budgetCategoriesId = BudgetCategoriesId(category.id, budget.id)
    }
}

@Embeddable
data class BudgetCategoriesId(
    @Column(name = "category_id")
    val categoryId: Int,
    @Column(name = "budget_id")
    val budgetId: Int
) : Serializable
